### PR TITLE
[JEP-237] introduce FIPS property

### PR DESCRIPTION
**Pull request by jtnord**
Introduces the code that core and plugins can use to switch behaviour based on FIPS-140 requirements

extracted from #8393 so it is easier to use in other PRs so that things are not help up waiting for a bigger PR.

See [JEP-237](https://github.com/jenkinsci/jep/tree/master/jep/237).

### Proposed changelog entries

- Developer: introduce JEP-237

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8482"></a>

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Review by olamy — APPROVED** [review]
LGTM
should we have a new label `fips` for triage and search?

**Comment by daniel-beck** [comment]
Have you considered how Jenkins should deal with incompatible new versions of the standard? Adding specific version (-2, -3, …) support for this API may allow plugins to disable features that would be compatible with older versions of the standard only. Perhaps that would be preferable to running noncompliant code that thinks it's compliant?

(Thinking out loud here, I see nothing wrong with this PR. "We'll deal with it when we get there" is a fine response.)

**Comment by jtnord** [comment]
> Have you considered how Jenkins should deal with incompatible new versions of the standard? Adding specific version (-2, -3, …) support for this API may allow plugins to disable features that would be compatible with older versions of the standard only. Perhaps that would be preferable to running noncompliant code that thinks it's compliant?

Mostly the difference is in the tightening of allowed hash algorithms, strengthening block ciphers (mostly use aes-128+) and digital signatures.

The place I expect this to impact plugins (or core) is more with the hashing algorithms, but where I have been looking for issues I have been checking code is both compliant with the intersection of FIPS-140-2 and FIPS-140-3.    This hasn't been explicitly called out - so calling this out is probably a good idea, although there is a lack of a good overview like we have from the bouncycastle user guide's approved sections.


> (Thinking out loud here, I see nothing wrong with this PR. "We'll deal with it when we get there" is a fine response.)

In many respects, yes if / when someone wants to move to FIPS-140-3 the code they wish to run would need to be checked again.  This is because the provider (which does not exist today) can implement a subset of the standards, so checking against the newer standard needs to really be checking against the newer provider that adheres to the standard.

**Comment by daniel-beck** [comment]
> Mostly the difference is in the tightening of allowed hash algorithms

Right, so you may have a plugin that implements a formerly compliant algorithm, but with too short a key length for the next version of the standard. The current API does not allow either core or the plugin itself to identify that what it does isn't OK with how Jenkins is expected to behave, and perhaps just refuse to do the thing, instead of being noncompliant.

Hence wondering whether a different API, e.g. an `EnumSet` getter of allowed FIPS versions, would allow callers to determine they're asked to implement FIPS but of a version they're not prepared for, and log this, or abort, or similar.

**Comment by jtnord** [comment]
> > Mostly the difference is in the tightening of allowed hash algorithms
> 
> Right, so you may have a plugin that implements a formerly compliant algorithm, but with too short a key length for the next version of the standard. The current API does not allow either core or the plugin itself to identify that what it does isn't OK with how Jenkins is expected to behave, and perhaps just refuse to do the thing, instead of being noncompliant.
> 
> Hence wondering whether a different API, e.g. an `EnumSet` getter of allowed FIPS versions, would allow callers to determine they're asked to implement FIPS but of a version they're not prepared for, and log this, or abort, or similar.

I was part way through editing my previous response but to keep a timeline :)

Update: I got confused and thought I was commenting on https://github.com/jenkins-infra/jenkins.io/pull/6675 for a moment.

As there is nothing stopping a plugin ignoring this flag - users need to make their own checks to satisfy themselves that the code/plugins they are using are compliant (or defer this checks to a 3rd party offering some support in this area 😁) so I do not immediately see an issue here.

I think that is probably the best place to capture this information?

> Right, so you may have a plugin that implements a formerly compliant algorithm, but with too short a key length for the next version of the standard. The current API does not allow either core or the plugin itself to identify that what it does isn't OK with how Jenkins is expected to behave, and perhaps just refuse to do the thing, instead of being noncompliant.

no different to having a plugin that does not even check this flag - there is nothing Jenkins can do here.  

If the algorithm is too short the JVM can (and should if configured correctly) tell the code to take a long walk of a short pier (ie throw some form of `Throwable`).  If it does not then it would appear that the JVM is not correctly configured.
The issue comes more with hashes than algorithms where use is nuanced..  You can not use MD5 for any security check (ie a HMac of a certificate, but you can use Md5 to validate the file you moved from A to B was not corrupted during the move - so long as the corruption not for preventing tampering - aka a security purpose)), so using MD5 itself would not blow up, but attempting to use it in a Mac should.

At this point the plugin should just be updated to use something present in both spec versions of the provider.

**Comment by daniel-beck** [comment]
> If the algorithm is too short the JVM can (and should if configured correctly) tell the code to take a long walk of a short pier (ie throw some form of `Throwable`).

I see; so any correctly implemented code after checking the property will throw in the provider. That's fine then 👍

**Review by daniel-beck — APPROVED** [review]

**Comment by jtnord** [comment]
> > If the algorithm is too short the JVM can (and should if configured correctly) tell the code to take a long walk of a short pier (ie throw some form of `Throwable`).
> 
> I see; so any correctly implemented code after checking the property will throw in the provider. That's fine then 👍

hopefully https://github.com/jenkins-infra/jenkins.io/blob/78d1678a3f1c92428f94db6764a4d301620d37ec/content/doc/book/system-administration/FIPS-140.adoc makes this clear.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
